### PR TITLE
feat: add support for multiple banners

### PR DESCRIPTION
**Pull request by aslilac**
Replace the singleton "service banner" with "notification banners" that can support any number of banners

Screenshot 2024-05-02 at 3 28 18 PM

- Migrate any existing banner to be a notification banner
- Rip out service banner logic/have it read from the new state depending on what makes sense
- Updated frontend for editing the notification banners state
- Remove a bunch of weird "preview" stuff, and just show the preview while you're editing

### CODE REVIEW

**Comment by aslilac** [comment]
TODO:

- [ ] I think the old database queries are safe to remove
- [ ] Add a db migration to move `service_banner` values to `notification_banners`
- [ ] Fix UI entitlement stuff

File: coderd/database/queries/siteconfig.sql [hunk 1]
```diff
@@ -43,6 +43,13 @@ ON CONFLICT (key) DO UPDATE SET value = $1 WHERE site_configs.key = 'service_ban
 -- name: GetServiceBanner :one
 SELECT value FROM site_configs WHERE key = 'service_banner';
 
+-- name: UpsertNotificationBanners :exec
+INSERT INTO site_configs (key, value) VALUES ('notification_banners', $1)
+ON CONFLICT (key) DO UPDATE SET value = $1 WHERE site_configs.key = 'notification_banners';
+
+-- name: GetNotificationBanners :one
+SELECT value FROM site_configs WHERE key = 'notification_banners';
```
Comment (Emyrk) [inline coderd/database/queries/siteconfig.sql:51]: I wonder if we should name the key like `notification_banners_json` to signal this is a json payload. I know `service_banner` is named without it. We can write a migration to remove `service_banner` entirely from the db.

File: codersdk/deployment.go [hunk 2]
```diff
@@ -2100,19 +2100,26 @@ func (c *Client) DeploymentStats(ctx context.Context) (DeploymentStats, error) {
...
+	// Deprecated: ServiceBanner is for a single banner, and has been replaced by NotificationBanners.
+	ServiceBanner       BannerConfig   `json:"service_banner"`
+	NotificationBanners []BannerConfig `json:"notification_banners"`
 }
 
-type ServiceBannerConfig struct {
+// ServiceBannerConfig has been renamed to BannerConfig.
+type ServiceBannerConfig = BannerConfig
+
+type BannerConfig struct {
```
Comment (Emyrk) [inline codersdk/deployment.go:2122]: Are all of these banners dismissable? Should we add some extra options here to control behavior?

File: coderd/database/queries/siteconfig.sql (hunk 1 above)
Comment (aslilac) [reply coderd/database/queries/siteconfig.sql:51]: There's a migration to convert any existing `'service_banner'` row into a `'notification_banners'` row. :) and I think the code makes it pretty clear that the row contains JSON

**Review by BrunoQuaresma — COMMENTED** [review]
There are a lot of things to review but I don't see anything concern in the FE code - I may have missed something - but during the QA I noticed a few things we could improve. Related to the BE, I'm going to rely on @Emyrk review.

**1. Make the table inside of the card full-width**
Usually, when tables are used inside of cards, they use a negative margin and remove the side and bottom borders to fill all the width.
Screenshot 2024-05-07 at 09 45 21

**2. Not allowing empty messages and use a default value instead of empty**
After clicking on "+New" and closing the dialog I saw a banner with an empty message in the table.

https://github.com/coder/coder/assets/3165839/cf2c4dee-ace1-485e-99fc-c93614e8ac30

**3. Message input is too large**
I have the impression that the input field is larger than it should be. If it's a textarea, we should make it clearer because it looks like a 2XL input.
Screenshot 2024-05-07 at 09 47 55

**Review by Emyrk — APPROVED** [review]
BE LG :+1:

File: agent/proto/agent.proto [hunk 3]
```diff
@@ -251,6 +251,18 @@ message BatchCreateLogsResponse {
...
+message GetNotificationBannersResponse {
+	repeated BannerConfig notification_banners = 1;
+}
+
+message BannerConfig {
+	bool enabled = 1;
+	string message = 2;
+	string background_color = 3;
+}
+
```
[the remark this answers is not in the document]
Comment (spikecurtis) [reply agent/proto/agent.proto:265]: Yes, minor bump for any new features if back compatible.

**Review by spikecurtis — CHANGES_REQUESTED** [review]
Needs a minor version bump

**Review by spikecurtis — APPROVED** [review]

**Review by BrunoQuaresma — APPROVED** [review]
Looks great! 

Just a nit but it is up to you. Having the banner getting added to the table when the dialog is open is a bit unexpected but anything that compromises the usage or experience.

**Pull request merged**